### PR TITLE
Implemented RelaxedOneHotCategorical + RelaxedBernoulli distributions

### PR DESCRIPTION
**Pull request by rachtsingh**
This implements the Gumbel-Softmax / Concrete distributions, which are continuous relaxations of discrete distributions, with reparametrizable gradients (i.e. for use in discrete VAEs).

Included are tests that:
1. Check that rounding samples from the `RelaxedBernoulli` gives samples from the corresponding `Bernoulli` distribution.
2. Check that taking the argmax of samples from `RelaxedOneHotCategorical` gives samples from the corresponding `OneHotCategorical` distribution.
3. Taking the `temperature -> \infty` gives a single point as samples.

Some math checked by @fritzo in https://github.com/probtorch/pytorch/pull/113

### CODE REVIEW

**Review by fritzo — APPROVED** [review]
LGTM

I've reviewed math for syntax errors and numerical stability, and I've reviewed tests to confirm they should detect any semantic errors.

**Comment by rachtsingh** [comment]
@apaszke anything else we need to do?

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Sorry for the delay, had a bunch of other PRs to review

**Comment by rachtsingh** [comment]
No worries, thanks for taking a look!
